### PR TITLE
Change nonsense return types to Nil: JSON and YAML

### PR DESCRIPTION
**Pull request by oprypin**
Strictly speaking, this is a backwards-incompatible change, but it should not affect any reasonable code. Attentive review is advised.

First I ran a tool to automatically insert return type restrictions based on what methods actually return in real usages. Then I picked out returns that are obviously purely accidental (due to implicitly returning the last expression) and can never be useful. So I change them to always be Nil.

**How to review this**:
The 1st commit exists only to provide a more useful diff in the 2nd commit.
**Please look only at the 2nd commit** (and not the overall diff) to see how this pull request changes actual return types that occur, into Nil.

### CODE REVIEW

**Comment by straight-shoota** [comment]
I suppose the builder methods could even return `self` to allow method chaining. But that should be a separate change.

**Comment by oprypin** [comment]
There is a bug here, the CI failure is real. Looking into it.
https://github.com/crystal-lang/crystal/pull/10622/checks

File: src/json/to_json.cr [hunk 1]
```diff
@@ -61,7 +61,7 @@ struct Float
 end
 
 class String
-  def to_json(json : JSON::Builder) : IO
+  def to_json(json : JSON::Builder) : Nil
```
Comment (oprypin) [inline src/json/to_json.cr:64]: Righto.... so it's this change that caused a bug to appear in the doc generator

Comment (straight-shoota) [reply src/json/to_json.cr:64]: Yeah, it's really just a bug in the doc generator.

https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/tools/doc/type.cr#L763

It should be `superclass ? superclass.to_json_simple(builder) || builder.null`

Comment (oprypin) [reply src/json/to_json.cr:64]: Right. So I pushed the fix.
It's a bit concerning that there's already a bug in stdlib itself, so what to say about other code...
On the other hand, maybe there aren't even such extensive usages of JSON::Builder

Comment (straight-shoota) [reply src/json/to_json.cr:64]: Well, at least this change helps identify such bugs.

**Pull request merged**